### ISSUE TITLE
Unable to compile MGARD

### ISSUE THREAD

**Issue by hengjiew** [CODARcode/MGARD#182]
Hello, I am trying to build MGARD and get the following error during compilation. 

```
/sw/summit/cuda/11.3.1/bin/nvcc -forward-unknown-to-host-compiler -DMGARD_ZSTD -Dmgard_library_EXPORTS -I/ccs/home/hengjie/packages/summit/MGARD/include -I/ccs/home/hengjie/packages/summit/MGARD/build/include -isystem=/sw/summit/python/3.8/anaconda3/2020.07-rhel8/include -isystem=/sw/summit/cuda/11.3.1/include -DMGARD_ENABLE_CUDA --expt-relaxed-constexpr -w -O3 -DNDEBUG --generate-code=arch=compute_70,code=[compute_70,sm_70] -Xcompiler=-fPIC -std=c++17 -MD -MT CMakeFiles/mgard-library.dir/src/mgard-x/DataRefactoring/CUDA/Decomposition_1D_Float.cu.o -MF CMakeFiles/mgard-library.dir/src/mgard-x/DataRefactoring/CUDA/Decomposition_1D_Float.cu.o.d -x cu -dc /ccs/home/hengjie/packages/summit/MGARD/src/mgard-x/DataRefactoring/CUDA/Decomposition_1D_Float.cu -o CMakeFiles/mgard-library.dir/src/mgard-x/DataRefactoring/CUDA/Decomposition_1D_Float.cu.o
/ccs/home/hengjie/packages/summit/MGARD/include/mgard-x/DataRefactoring/MultiDimension/DataRefactoring.hpp: In function 'void mgard_x::decompose(mgard_x::Hierarchy<D, T, DeviceType>&, mgard_x::SubArray<D, T, DeviceType>&, mgard_x::SIZE, int)':
/ccs/home/hengjie/packages/summit/MGARD/include/mgard-x/DataRefactoring/MultiDimension/DataRefactoring.hpp:1713:30: error: 'auto' parameter not permitted in this context
 1713 |   SubArray w(workspace);
      |                              ^
/ccs/home/hengjie/packages/summit/MGARD/include/mgard-x/DataRefactoring/MultiDimension/DataRefactoring.hpp:1713:10: error: deduced class type 'SubArray' in function return type
 1713 |   SubArray w(workspace);
      |          ^
/ccs/home/hengjie/packages/summit/MGARD/include/mgard-x/DataRefactoring/MultiDimension/../../RuntimeX/DeviceAdapters/../DataStructures/SubArray.hpp:16:51: note: 'template<unsigned int D, class T, class DeviceType> class mgard_x::SubArray' declared here
   16 | template <DIM D, typename T, typename DeviceType> class SubArray {
      |                                                   ^~~~~~~~
/ccs/home/hengjie/packages/summit/MGARD/include/mgard-x/DataRefactoring/MultiDimension/DataRefactoring.hpp:1722:131: error: 'w' was not declared in this scope
 1722 |       LwpkReo<D, T, COPY, DeviceType>().Execute(
```

I have installed MGARD's dependencies and tried to build it with cuda 11.3.1 as follow:
```
cmake -S . -B build -DMGARD_ENABLE_CUDA=ON -DCMAKE_CUDA_ARCHITECTURES=70 -DCMAKE_CXX_STANDARD=17
```
Then I compiled MGARD in the build directory via `make -j VERBOSE=1` and saw the above error. I had the same error without `CMAKE_CXX_STANDARD=17` and added it to ensure that `auto` is supported. But it didn't help.

Am I missing some setups during the build? 
Thanks a lot for your help!

BTW, the one-step configure and build script in MGARD-X' readme page links to an invalid webpage.

**Comment by JieyangChen7** [comment]
@hengjiew Thanks for pointing out the error on our readme page. We will fix the link.

I'm able to reproduce this error with CUDA v11.3.1 on Summit. However, the error disappears if I use a newer version (e.g., 11.4) or an older version (e.g., 11.1). So, it seems that this is due to a bug in the NVCC compiler. Could you try to use a different version of CUDA?
You can also use the build script we wrote specifically for Summit: https://github.com/CODARcode/MGARD/blob/master/build_scripts/build_mgard_cuda_summit.sh

Thanks!

**Comment by hengjiew** [comment]
> 

Hello @JieyangChen7 , thanks a lot for looking into this issue. 
I can compile MGARD with cuda/11.4.2 on Summit. 

I also tried to run the script `build_mgard_cuda_summit.h` but it gave me the following error:
```
In file included from /ccs/home/hengjie/packages/summit/MGARD/build-cuda-summit/mgard/include/proto/mgard.pb.h:26,
                 from /ccs/home/hengjie/packages/summit/MGARD/include/format.hpp:12,
                 from /ccs/home/hengjie/packages/summit/MGARD/src/format.cpp:1:
/ccs/home/hengjie/packages/summit/MGARD/install-cuda-summit/include/google/protobuf/generated_message_table_driven.h:299:25: error: 'MapEntryHelper' does not name a type; did you mean 'MapEntryLite'?
  299 |   bool operator()(const MapEntryHelper<T>& a,
      |                         ^~~~~~~~~~~~~~
      |                         MapEntryLite
compilation terminated due to -Wfatal-errors.
```

**Comment by JieyangChen7** [comment]
@ben-e-whitney This seems to be an error when calling Protobuf. Have you seen this before? I have never encountered this issue on Summit.

**Comment by ben-e-whitney** [comment]
I haven't, no. @hengjiew, which C++ compiler are you using?

**Comment by hengjiew** [comment]
> I haven't, no. @hengjiew, which C++ compiler are you using?

Hi @ben-e-whitney, I am using gcc/9.3.0.

**Comment by ben-e-whitney** [comment]
I am having trouble reproducing this error. It appears to be related only to Protobuf, not CUDA, so I've made a simplified build script that enables only the serial MGARD-X backend. I've pasted the script below. It runs without any errors for me. @hengjiew, please try running it and tell us what happens.

```sh
#!/bin/sh

set -e

module purge
module load gcc/9.3.0
module load zstd/1.5.0
module load cmake/3.21.3

set -x

######## User Configurations ########
# Source directory
mgard_src_dir=.
# Build directory
build_dir=./build-summit-simplified
# Number of processors used for building
num_build_procs=8
# Installtaion directory
install_dir=./install-summit-simplified

mkdir -p "${build_dir}"

#build Protobuf
protobuf_dir=${build_dir}/protobuf-3.19.4
protobuf_src_dir=${protobuf_dir}/cmake
protobuf_build_dir=${protobuf_dir}/build
protobuf_install_dir=${install_dir}
if [ ! -d "${protobuf_dir}" ]; then
	pushd "${build_dir}"
	wget "https://github.com/protocolbuffers/protobuf/releases/download/v3.19.4/protobuf-cpp-3.19.4.tar.gz"
	gunzip "protobuf-cpp-3.19.4.tar.gz"
	tar --file "protobuf-cpp-3.19.4.tar" --extract
	popd
fi
cmake -S ${protobuf_src_dir} -B ${protobuf_build_dir}\
    -Dprotobuf_BUILD_SHARED_LIBS=ON\
    -DCMAKE_INSTALL_PREFIX=${protobuf_install_dir}
cmake --build ${protobuf_build_dir} -j ${num_build_procs}
cmake --install ${protobuf_build_dir}

#build MGARD
mgard_build_dir=${build_dir}/mgard
mgard_install_dir=${install_dir}
mkdir -p ${mgard_build_dir}
cmake -S ${mgard_src_dir} -B ${mgard_build_dir} \
    -DCMAKE_PREFIX_PATH="${protobuf_install_dir}"\
    -DMGARD_ENABLE_SERIAL=ON\
    -DBUILD_TESTING=OFF\
    -DCMAKE_BUILD_TYPE=Release\
    -DCMAKE_INSTALL_PREFIX=${mgard_install_dir}
cmake --build ${mgard_build_dir} -j ${num_build_procs}
cmake --install ${mgard_build_dir}
```

**Comment by hengjiew** [comment]
Hello @ben-e-whitney, thanks for looking into this. 
I got the same error with your script and attached the build log. 
[build_err.log](https://github.com/CODARcode/MGARD/files/8226379/build_err.log)

**Comment by ben-e-whitney** [comment]
@hengjiew According to the log, you're building Protobuf with `gcc` 8.3.1 but MGARD with `gcc` 9.3.0. Can you try using version 9.3.0 for both?

**Comment by hengjiew** [comment]
> @hengjiew According to the log, you're building Protobuf with `gcc` 8.3.1 but MGARD with `gcc` 9.3.0. Can you try using version 9.3.0 for both?
> 
> (I'm not sure why CMake is finding one compiler when it configures the Protobuf build and another when it configures the MGARD build. Maybe try manually specifying the compiler with `-D CMAKE_C_COMPILER="$(which gcc)" -D CMAKE_CXX_COMPILER="$(which g++)"`.)

@ben-e-whitney I was able to bluid MGARD when I installed all the dependencies manually with the same compiler (gcc/9.3.0). So I think you are right that using different compiler causes the failure for the build script. Thanks!

**Comment by ben-e-whitney** [comment]
I believe that everything raised in this issue will be resolved once #188 is merged (@JieyangChen7 fixed the broken link there). @hengjiew, please let me know if I'm mistaken or if you've encountered any new problems building MGARD.
